File: OneCodeTeam/How-to-use-App-Specific-Hardware-ID-ASHWID-to-22437/[VB]-How to use App Specific Hardware ID (ASHWID) to implement per-device logic in Wi/VB/VBWindowsStoreAppASHWID/VBWindowsStoreAppDeviceClient/ViewModels/DelegateCommand.vb
'************************************* Module Header ***********************\
' Module Name:  DelegateCommand.vb
' Project:      VBWindowsStoreAppASHWID
' Copyright (c) Microsoft Corporation.
' 
' Helper class to create a quick ICommand object.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/en-us/openness/resources/licenses.aspx#MPL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'*****************************************************************************/

Imports System.Windows.Input

''' <summary>
''' Helper class to create a quick ICommand object.
''' </summary>
''' <remarks></remarks>
Public Class DelegateCommand
    Implements ICommand

    Private ReadOnly _canExecute As Predicate(Of Object)
    Private ReadOnly _execute As Action(Of Object)

    Public Sub New(ByVal method As Action(Of Object))
        Me.New(method, Nothing)
    End Sub

    Public Sub New(ByVal execute As Action(Of Object), ByVal canExecute As Predicate(Of Object))
        _execute = execute
        _canExecute = canExecute
    End Sub

    Public Function CanExecute(ByVal parameter As Object) As Boolean Implements ICommand.CanExecute
        Return (_canExecute Is Nothing OrElse _canExecute(parameter))
    End Function

    Public Sub Execute(ByVal parameter As Object) Implements ICommand.Execute
        If (Not _execute Is Nothing) Then
            _execute(parameter)
        End If
    End Sub

    Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

    Protected Overridable Sub OnCanExecuteChanged(ByVal e As EventArgs)
        RaiseEvent CanExecuteChanged(Me, e)
    End Sub

    Public Sub RaiseCanExecuteChanged()
        OnCanExecuteChanged(EventArgs.Empty)
    End Sub

End Class

File: OneCodeTeam/How-to-use-ASHWID-to-implement-per-device-logic-21358/[C#]-How to use ASHWID to implement per-device logic safely/C#/CSWindowsStoreAppASHWID/CSWindowsStoreAppDeviceClient/ViewModel/DelegateCommand.cs
/****************************** Module Header ******************************\
 * Module Name:  DelegateCommand.cs
 * Project:      CSWindowsStoreAppDeviceClient
 * Copyright (c) Microsoft Corporation.
 * 
 * Helper class to create a quick ICommand object.
 *  
 * This source is subject to the Microsoft Public License.
 * See http://www.microsoft.com/en-us/openness/licenses.aspx#MPL
 * All other rights reserved.
 * 
 * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
 * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
 * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

using System;
using System.Windows.Input;

namespace CSWindowsStoreAppDeviceClient.ViewModel
{
    /// <summary>
    /// A ICommand helper whose delegates can be attached
    /// </summary>
    class DelegateCommand : ICommand
    {
        private readonly Predicate<object> _canExecute;
        private readonly Action<object> _execute;

        public DelegateCommand(Action<object> method) : this(method, null) {}

        public DelegateCommand(Action<object> execute, Predicate<object> canExecute)
        {
            _execute = execute;
            _canExecute = canExecute;
        }

        public bool CanExecute(object parameter)
        {
            return _canExecute == null || _canExecute(parameter);
        }

        public void Execute(object parameter)
        {
            if (_execute != null)
            {
                _execute(parameter);
            }
        }

        public event EventHandler CanExecuteChanged;

        protected virtual void OnCanExecuteChanged(EventArgs e)
        {
            if (CanExecuteChanged != null)
            {
                CanExecuteChanged(this, e);
            }
        }

        public void RaiseCanExecuteChanged()
        {
            OnCanExecuteChanged(EventArgs.Empty);
        }
    }
}
